[PyTorch Edge] Fix setting `training` attribute logic`

cccclai opened the pull request:
Update previous recursive logic. 
1. Continue setting training attribute only if the slot is an object and a module.
2. Remove the TORCH_INTERNAL_ASSERT. If there is no `training` attribute, we just don't set it. For the corresponding JIT module, they get the module list first and set module one by one. there is method to get all modules iteratively, instead of recursively. When the current slot is a module, they have a different way to scan recursively.
> if the current thing is a module, we have to scan it for recursive traversals.




Test:
```
TEST(LiteTrainerTest, CustomOP) {
  std::string model_file =
      "/Users/chenlai/Documents/pytorch/data/model_f269583363.ptl";
  Module m = load(model_file);
  m.train(true);

  mobile::Module mobile_module = _load_for_mobile(model_file);
  mobile_module.train(true);
}
```

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59180 [PyTorch Edge] Fix setting `training` attribute logic`**

Differential Revision: [D28779374](https://our.internmc.facebook.com/intern/diff/D28779374)

cccclai committed:
[PyTorch Edge] Fix setting `training` attribute logic`

[ghstack-poisoned]

cccclai commented:
@cccclai has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28779374).

cccclai committed:
Update on "[PyTorch Edge] Fix setting `training` attribute logic`"


Differential Revision: [D28779374](https://our.internmc.facebook.com/intern/diff/D28779374)

[ghstack-poisoned]

cccclai commented:
@cccclai has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28779374).

iseeyuan commented on torch/csrc/jit/mobile/module.cpp:
```diff
@@ -46,14 +48,18 @@ namespace {
-      set_train_recurse(slot.toObject(), on);
+    if (slot.isObject() && slot.toObjectRef().type()->is_module()) {
+      auto slot_obj = slot.toObject();
+      if (slot_obj->type()->hasAttribute("training")) {
```
Is it possible that a module Parent does not have "training", but it's submodule Child has "training". Would it stop here at Parent and miss set_train for Child?

cccclai replied:
Modify the logic in the latest commit. One thing I don't understand is the `TORCH_INTERNAL_ASSERT` part. Do we expect all modules have this attribute?

cccclai committed:
Update on "[PyTorch Edge] Fix setting `training` attribute logic`"


Differential Revision: [D28779374](https://our.internmc.facebook.com/intern/diff/D28779374)

[ghstack-poisoned]

cccclai commented:
@cccclai has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28779374).

cccclai committed:
Update on "[PyTorch Edge] Fix setting `training` attribute logic`"


Differential Revision: [D28779374](https://our.internmc.facebook.com/intern/diff/D28779374)

[ghstack-poisoned]

cccclai committed:
Update on "[PyTorch Edge] Fix setting `training` attribute logic`"


Differential Revision: [D28779374](https://our.internmc.facebook.com/intern/diff/D28779374)

[ghstack-poisoned]

cccclai committed:
Update on "[PyTorch Edge] Fix setting `training` attribute logic`"


Differential Revision: [D28779374](https://our.internmc.facebook.com/intern/diff/D28779374)

[ghstack-poisoned]

cccclai committed:
Update on "[PyTorch Edge] Fix setting `training` attribute logic`"


Differential Revision: [D28779374](https://our.internmc.facebook.com/intern/diff/D28779374)

[ghstack-poisoned]

cccclai commented:
@cccclai has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28779374).

cccclai committed:
Update on "[PyTorch Edge] Fix setting `training` attribute logic`"

Update previous recursive logic. 
1. Continue setting training attribute only if the slot is an object and a module.
2. Remove the TORCH_INTERNAL_ASSERT. If there is no `training` attribute, we just don't set it. For the corresponding JIT module, they get the module list first and set module one by one. there is method to get all modules iteratively, instead of recursively. When the current slot is a module, they have a different way to scan recursively.
> if the current thing is a module, we have to scan it for recursive traversals.




Test:
```
TEST(LiteTrainerTest, CustomOP) {
  std::string model_file =
      "/Users/chenlai/Documents/pytorch/data/model_f269583363.ptl";
  Module m = load(model_file);
  m.train(true);

  mobile::Module mobile_module = _load_for_mobile(model_file);
  mobile_module.train(true);
}
```


Differential Revision: [D28779374](https://our.internmc.facebook.com/intern/diff/D28779374)

[ghstack-poisoned]

cccclai commented:
@cccclai has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28779374).

iseeyuan commented on torch/csrc/jit/mobile/module.cpp:
```diff
@@ -43,16 +43,20 @@ c10::optional<Method> Module::find_method(const std::string& basename) const {
-    if (slot.isObject()) {
+    // slots is a list of IValue. Continue setting training attribute only
+    // if the slot is an Object and a module.
+    if (slot.isObject() && slot.toObjectRef().type()->is_module()) {
```
I'm not sure. Object holds slots. What happens if an object is not a module, but one of its slots is a module? Is it possible?

iseeyuan commented on torch/csrc/jit/mobile/module.cpp:
```diff
@@ -43,16 +43,20 @@ c10::optional<Method> Module::find_method(const std::string& basename) const {
 namespace {
+
+// For JIT, there is a function to get all modules by iteration in struct slot_iterator_impl
+// (jit/api/module.h), this function use recursion to mimic the logic without allocating extra
```
Could you also provide the code location of how jit set the training property?

iseeyuan reviewed (commented):
Could you add the unit test to this PR? It would be great to extract the property of the model file and construct the smaller model in the test.

cccclai committed:
Update on "[PyTorch Edge] Fix setting `training` attribute logic`"

Update previous recursive logic. 
1. Continue setting training attribute only if the slot is an object and a module.
2. Remove the TORCH_INTERNAL_ASSERT. If there is no `training` attribute, we just don't set it. For the corresponding JIT module, they get the module list first and set module one by one. there is method to get all modules iteratively, instead of recursively. When the current slot is a module, they have a different way to scan recursively.
> if the current thing is a module, we have to scan it for recursive traversals.




Test:
```
TEST(LiteTrainerTest, CustomOP) {
  std::string model_file =
      "/Users/chenlai/Documents/pytorch/data/model_f269583363.ptl";
  Module m = load(model_file);
  m.train(true);

  mobile::Module mobile_module = _load_for_mobile(model_file);
  mobile_module.train(true);
}
```


Differential Revision: [D28779374](https://our.internmc.facebook.com/intern/diff/D28779374)

[ghstack-poisoned]

cccclai committed:
Update on "[PyTorch Edge] Fix setting `training` attribute logic`"

Update previous recursive logic. 
1. Continue setting training attribute only if the slot is an object and a module.
2. Remove the TORCH_INTERNAL_ASSERT. If there is no `training` attribute, we just don't set it. For the corresponding JIT module, they get the module list first and set module one by one. there is method to get all modules iteratively, instead of recursively. When the current slot is a module, they have a different way to scan recursively.
> if the current thing is a module, we have to scan it for recursive traversals.




Test:
```
TEST(LiteTrainerTest, CustomOP) {
  std::string model_file =
      "/Users/chenlai/Documents/pytorch/data/model_f269583363.ptl";
  Module m = load(model_file);
  m.train(true);

  mobile::Module mobile_module = _load_for_mobile(model_file);
  mobile_module.train(true);
}
```


Differential Revision: [D28779374](https://our.internmc.facebook.com/intern/diff/D28779374)

[ghstack-poisoned]
